core/src/test/java/org/apache/druid/data/input/HandlingInputRowIteratorTest.java

```diff
@@ -0,0 +1,183 @@
+    private static final List<InputRow> INPUT_ROWS = Arrays.asList(INPUT_ROW1, INPUT_ROW2);
+
+    private TestInputRowHandler successfulHandler;
+    private TestInputRowHandler unsuccessfulHandler;
+
+    @Before
+    public void setup()
+    {
+      successfulHandler = new TestInputRowHandler(true);
+      unsuccessfulHandler = new TestInputRowHandler(false);
+    }
+
+    @Test
+    public void hasNext()
+    {
+      HandlingInputRowIterator target = createInputRowIterator(unsuccessfulHandler, unsuccessfulHandler);
+      Assert.assertTrue(target.hasNext());
+      Assert.assertFalse(unsuccessfulHandler.invoked);
+    }
+
+    @Test
+    public void yieldsNextIfUnhandled()
+    {
+      HandlingInputRowIterator target = createInputRowIterator(unsuccessfulHandler, unsuccessfulHandler);
+      Assert.assertEquals(INPUT_ROW1, target.next());
+      Assert.assertTrue(unsuccessfulHandler.invoked);
+    }
+
+    @Test
+    public void yieldsNullIfHandledByFirst()
+    {
+      HandlingInputRowIterator target = createInputRowIterator(successfulHandler, unsuccessfulHandler);
+      Assert.assertNull(target.next());
+      Assert.assertTrue(successfulHandler.invoked);
+      Assert.assertFalse(unsuccessfulHandler.invoked);
+    }
+
+    @Test
+    public void yieldsNullIfHandledBySecond()
+    {
+      HandlingInputRowIterator target = createInputRowIterator(unsuccessfulHandler, successfulHandler);
+      Assert.assertNull(target.next());
+      Assert.assertTrue(unsuccessfulHandler.invoked);
+      Assert.assertTrue(successfulHandler.invoked);
+    }
+
+    private static HandlingInputRowIterator createInputRowIterator(
+        HandlingInputRowIterator.InputRowHandler firstHandler,
+        HandlingInputRowIterator.InputRowHandler secondHandler
+    )
+    {
+      CloseableIterator<InputRow> iterator = new TestCloseableIterator()
+      {
+        private final Iterator<InputRow> delegate = INPUT_ROWS.iterator();
+
+        @Override
+        public boolean hasNext()
+        {
+          return delegate.hasNext();
+        }
+
+        @Nullable
+        @Override
+        public InputRow next()
+        {
+          return delegate.next();
+        }
+      };
+
+      return new HandlingInputRowIterator(iterator, Arrays.asList(firstHandler, secondHandler));
+    }
+
+    private static class TestInputRowHandler implements HandlingInputRowIterator.InputRowHandler
+    {
+      boolean invoked = false;
+
+      private final boolean successful;
+
+      TestInputRowHandler(boolean successful)
+      {
+        this.successful = successful;
+      }
+
+      @Override
+      public boolean handle(InputRow inputRow)
+      {
+        invoked = true;
+        return successful;
+      }
+    }
+  }
+
+  private abstract static class TestCloseableIterator implements CloseableIterator<InputRow>
+  {
+    @Override
+    public void close()
+    {
+      throw new UnsupportedOperationException();
```

Comment:
It is for the convenience of not having to override `close()` when creating each of the iterators that are used for the tests.

Reply:
Hmm, I think it will be more clear to not extend `CloseableIterator` in that case. If it had to extend `CloseableIterator`, it makes more sense to me to call `close()` properly.

Reply:
Changed this to use `CloseableIterators.withEmptyBaggage()`